feat(wc): duplicate orders admin notice

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Adds an admin notice when duplicate orders are detected:

<img width="1165" alt="image" src="https://github.com/user-attachments/assets/49de1953-30b0-43d1-8a6e-9d76c91b9b14">

### How to test the changes in this Pull Request:

1. Donate two times with the same amount, within 10 minutes
2. Do that again, but change the dates of these two to more than a month ago
3. Trigger the CRON job: `wp cron event run newspack_wc_check_order_duplicates`
4. Visit the admin panel, observe the admin notice about potentially duplicated orders – from the last month (the first two created)
5. Dismiss one of the cases, observe the page reloads and the list item is no longer displayed
6. Run `wp newspack detect-order-duplicates --cutoff_time='1 year' --save` and refresh the admin page
7. Observe three items are present on the list: two new ones from more than a month ago

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208711277186607

adekbadek committed:
feat(wc): duplicate orders admin notice

adekbadek committed:
feat: handle HPOS

adekbadek committed:
feat: handle order series dismissal

adekbadek committed:
feat: better UX with a details element

leogermani reviewed (commented):
Even though the SQL queries are pretty efficient, they are huge, hard to read and hard to maintain.

And also hard to test, because I need two different sites to test it.

Doing this with simple code, looping though recent orders would be less efficient, but much more readable, maintainable and testable... And considering that this check runs daily, we don't need to look at all the orders in the history, just the most recent ones.

adekbadek committed:
feat: iterate orders instead of using a direct DB query

adekbadek committed:
chore: unify wording

adekbadek committed:
fix: handle the order meta

adekbadek committed:
feat: CLI tool

adekbadek committed:
feat: process orders in batches

adekbadek commented:
Yeah, that makes sense. I've updated with the following:

1. Instead of SQL query, iterate order objects retrieved with `wc_get_orders` (in batches of 100, so the site does not fall over)
2. Added a `wp newspack detect-order-duplicates` CLI command to enable filling in historical data (older than a month). This will have to be ran on live sites after the PR is deployed.

Update the PR description accordingly.

leogermani commented on includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php:
```diff
@@ -0,0 +1,247 @@
+	}
+
+	/**
+	 * CLI handler to upsert the order duplicates with a specified timeframe.
```
is it detect or upsert?

leogermani replied:
Let's add a clearer description of what the command does
- search for duplicates and optionally store this info to be displayed in the admin panel

adekbadek replied:
Updated in 520f49c1367f00aebf272734c855943de2c7080a

leogermani commented on includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php:
```diff
@@ -0,0 +1,247 @@
+	 *
+	 * ## EXAMPLES
+	 *
+	 *     wp newspack detect-order-duplicates --cutoff_time='2 months' --save
```
is it detect or upsert?

adekbadek replied:
Without `--save`, it's only detection. With it, it's upsertion.

leogermani reviewed (commented):
Works as described! Thanks for the changes! I left a few minor comments.

I just noticed one thing. I first ran the CLI command without the `--save` param. It detected the duplicated but didn't save. However, it ended up displaying the admin notice, but with no orders listed. Just the summary. When you clicked, it was an empty list.

Another question: Since we're running this daily, and cutting off based on completion date, why default to check the last 30 days? Looks like we will be doing a lot of unnecessary checks every day. Isn't it enough to check just the last one or two days?

adekbadek committed:
feat: tweak

adekbadek commented:
> However, it ended up displaying the admin notice, but with no orders listed.

Yeah, I failed to account for an empty state, thanks for catching that! Fixed in 520f49c1367f00aebf272734c855943de2c7080a

> Isn't it enough to check just the last one or two days?

That's right, changed to a day in 520f49c1367f00aebf272734c855943de2c7080a
